### PR TITLE
Skips DDP parameter sync

### LINKED ISSUE #2595: ddp no_sync
## 🚀 Feature
DDP has this no_sync context manager.
https://pytorch.org/docs/stable/nn.html?highlight=no_sync#torch.nn.parallel.DistributedDataParallel.no_sync

Could be good to add it when doing gradient accumulation.

### LINKED ISSUE #4092: Avoid unnecessary DDP synchronization when gradient_accumulation_steps > 1
## 🚀 Feature
Avoid unnecessary DDP synchronization when gradient_accumulation_steps > 1 

### Motivation

When training large models the synchronization  is costly and the actual speedup from 2 gpus is much lower than 200%

### Pitch

We can use DDP [no_sync](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync) feature to avoid synchronization in steps that doesn't call ```optimizer_step```

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Skips DDP parameter sync  in forward and backward whenever possible.

Fixes #4092 and fixes #2595 

Is a revamp of #4146 after adding optimiser closures.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by williamFalcon** [comment]
cc @ananthsub for review

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -679,20 +681,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+                    # no ddp sync at the beginning of forward or backward due to parameter changes in this or last step required
+                    no_sync = self._updated_model_last_step and isinstance(self.trainer.model, torch.nn.parallel.DistributedDataParallel)
+                    if no_sync:
+                        self.trainer.model.no_sync.__enter__()
+
                     self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
+                    if no_sync:
+                        self.trainer.model.__exit__()
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:694]: is it possible for this to logic to live in the DDP accelerators as opposed to the general training loop?

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -679,20 +681,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+                    # no ddp sync at the beginning of forward or backward due to parameter changes in this or last step required
+                    no_sync = self._updated_model_last_step and isinstance(self.trainer.model, torch.nn.parallel.DistributedDataParallel)
+                    if no_sync:
+                        self.trainer.model.no_sync.__enter__()
+
                     self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
+                    if no_sync:
+                        self.trainer.model.__exit__()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:693]:
```suggestion
                        self.trainer.model.no_sync.__exit__()
```

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from copy import copy, deepcopy
 
 import numpy as np
+from numpy.lib.arraysetops import isin
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:19]: where's this used?

**Comment by awaelchli** [comment]
This is implementing the same as requested [here](#2595), right?

**Review by ananthsub — APPROVED** [review]
besides my small comments, the PR looks good to me. the accelerator part is my only significant question. if it's not immediately obvious how we can move to this check to the DDP accelerators, then let's land this to reap the perf savings and figure out the refactor later. i really like that the training loop now barely has any references to pytorch specifics

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/trainer/training_loop.py:19]: it's not. was added by my IDE as auto import during typing `isinstance` :D
removed it

**Comment by justusschock** [comment]
@awaelchli yes, haven't seen that one.

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -679,20 +680,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
                     # -------------------
                     # calculate loss (train step + train step end)
                     # -------------------
+
+                    # no ddp sync at the beginning of forward or backward due to parameter changes in this or last step required
+                    no_sync = self._updated_model_last_step and isinstance(self.trainer.model, torch.nn.parallel.DistributedDataParallel)
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:685]: Great catch there !

Would you mind making this a bit more readable ? 
Like moving no_sync outside of main loop and hiding those __enter__() and __exit__() ?
Also, you don't need `self._updated_model_last_step` variable. It can derived from `self.trainer.batch_idx and self.trainer.accumulate_grad_batches`

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/trainer/training_loop.py:688]: @tchaton Regarding the variable you're probably right. 

Where do you want to move it and how do you want to hide those enter and exit methods?
I haven't found a better place for it, since it affects every DDP backend we have so far.

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -676,26 +678,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
 
                     # -------------------
                     # calculate loss (train step + train step end)
                     # -------------------
-                    self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
+                    # perform dpp sync only when performing optimizer_step
+                    with self.block_ddp_sync_behaviour():
+                        self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:691]: @justusschock @awaelchli @williamFalcon I think the `backward` in the DDP accelerator should use the wrapped DDP model, not the module inside. Otherwise `require_backward_grad_sync=False` on the wrapped DDP model is being ignored and we miss the grad accumulation benefits

https://github.com/PyTorchLightning/pytorch-lightning/blob/41de4538aa0c187793709a93875e67666c2ddde8/pytorch_lightning/accelerators/accelerator.py#L89-L101

https://github.com/PyTorchLightning/pytorch-lightning/blob/41de4538aa0c187793709a93875e67666c2ddde8/pytorch_lightning/trainer/connectors/model_connector.py#L54-L57

Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:691]: https://github.com/PyTorchLightning/pytorch-lightning/pull/4415

Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:691]: do we need `block_ddp_sync_behaviour` ? 
```suggestion
                    # perform dpp sync only when performing optimizer_step
                    if isinstance(self.trainer.model, torch.nn.parallel.DistributedDataParallel):
                        with self.trainer.model.no_sync():
                            self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
                     else:
                          self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
                    with self.block_ddp_sync_behaviour():
                        self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
```

File: pytorch_lightning/trainer/training_loop.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ def __init__(self, trainer):
         self.automatic_optimization = True
         self._curr_step_result = None
         self._cur_grad_norm_dict = None
+        self._updated_model_last_step = False
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:49]: Can you please remove `_updated_model_last_step `. You don't use it anymore :)

File: pytorch_lightning/trainer/training_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:691]: Just to make it more explicit for our readers :)

**Review by tchaton — CHANGES_REQUESTED** [review]
Great catch !

File: pytorch_lightning/trainer/training_loop.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/trainer/training_loop.py:690]: @tchaton requested this to make it more readable and hide the conditions for that in the context manager

**Review by tchaton — APPROVED** [review]
Great addition !

File: pytorch_lightning/trainer/training_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:689]: @ananthsub 
The training and evaluation loop should look absolutely perfect and simple to understand.
As a new coder, I should recognise my training loop. As a new coder, I have no knowledge about ddp and your suggested change would have confused me :)

One great example of this PR is: not (accumulation_done or is_final_batch) -> should_accumulate. It is pretty simple, but make a clear statement about what is happening. We should try to enforce this as much as possible :)

I hope it makes sense :)

Comment (SeanNaren) [reply pytorch_lightning/trainer/training_loop.py:689]: This was resolved after realising LightningDistributedDataParallel/LightningDataParallel do not override the BW pass

**Review by s-rog — APPROVED** [review]
Awesome! Should be a nice performance boost!

**Review by rohitgr7 — APPROVED** [review]
